### ISSUE TITLE
Slower than new PyTorch GPU transformations

### ISSUE THREAD

**Issue by FrancescoSaverioZuppichini** [kornia/kornia#966]
## ❓ Questions and Help
Kornia's augmentations seem to be slower than PyTorch's one. The dataset I am using is CelebA.

I have:

OS: Ubuntu 20.04
GPU: 1080ti
PyTorch: 1.8.0
torchvision: 0.9.0

```python

from torchvision.io import read_image
from torchvision.datasets import ImageFolder
from torch.utils.data import DataLoader
import torchvision.transforms as TT
import kornia.augmentation as KT

def timeit(func):
    def _inner(*args, **kwargs):
        start = time()
        func(*args, **kwargs)
        end = time()
        print(f"Elapsed = {end - start :.2f}s")

    return _inner

@timeit
def test_augmentation(tr, device="cuda"):
    ds = ImageFolder(root="~/Documents/Datasets/celeba", loader=read_image)
    dl = DataLoader(ds, batch_size=128, num_workers=16, pin_memory=True)

    for batch in tqdm(dl):
        x = batch[0].to(device)
        tr(x)
        continue


def test_augmentations():
    resize = 256
    crop = 224
    device = "cuda"

    augmentations = [
        nn.Sequential(
            TT.Resize((resize, resize)),
            TT.ConvertImageDtype(dtype=torch.float),
            TT.RandomVerticalFlip(),
            TT.RandomHorizontalFlip(),
            TT.RandomCrop(crop),
        ).to(device),
        nn.Sequential(
            TT.Resize((resize, resize)),
            TT.ConvertImageDtype(dtype=torch.float),
            KT.RandomVerticalFlip(),
            KT.RandomHorizontalFlip(),
            KT.RandomCrop((crop, crop)),
        ).to(device),
    ]

    list(map(test_augmentation, augmentations))


test_augmentations()
```
PyTorch's augmentations takes 15s while Kornia takes 29s. 

Best regards,

**Comment by shijianjian** [comment]
Hi @FrancescoSaverioZuppichini,

Thank you for reporting. Torchvision implemented the RandomCrop and related in a non-differentiable way, while Kornia is not. 

We are aware of the timing issue that brought by the differentiable cropping implementation, and we are planning on incorporating the non-differentiable implementation for those common usages. 

Best,
Jian

**Comment by edgarriba** [comment]
besides what @shijianjian pointed about  differentiability, kornia is meant to work better in batch on device which means adding kornia.augmentation as a layer before your model instead of inside the dataloader.

**Comment by FrancescoSaverioZuppichini** [comment]
Hi @shijianjian, thank you for the quick reply. 

> Besides what @shijianjian pointed about differentiability, kornia is meant to work better in batch on device which means adding kornia.augmentation as a layer before your model instead of inside the dataloader.

That's the same thing, It won't change performances. I am running them on the batched data on the GPU.

I am felling pretty dumb today, but correct me if I am wrong. The following snippet:

```python
import torchvision.transforms as TT

resize = 256
crop = 224
device = "cuda"
x = torch.randn((1, 3, 224, 224), requires_grad=True)
y = torch.ones_like(x)
# torchvision augmentation
tr = nn.Sequential(TT.RandomVerticalFlip())

model = nn.Conv2d(3, 3, kernel_size=1)

criterion = nn.MSELoss()
optimizer = opt.Adam(model.parameters())
for _ in range(100):
    out = tr(model(x))
    loss = criterion(y, out)
    optimizer.zero_grad()
    loss.backward()
    print(loss)
    optimizer.step()
```

Shouldn't work since `torchvision.augmentation` is not differentiable, but it works. I think I am missing something. Nevertheless, thank you!

**Comment by shijianjian** [comment]
Non-differentiable does not necessarily mean to break the computational graph. As a differentiable operation, apart from retaining the computational graph, it will also provide meaningful gradients. e.g. the crop function in kornia is based on perspective transform which provides the gradients w.r.t. the grid.

In your example, the gradients will be retained since the graph is still intact. Under the hood, TorchVision and Kornia implemented the flip in the same way. For cropping and related, kornia will provide traceable gradients that differs from torchvision.

**Comment by FrancescoSaverioZuppichini** [comment]
Thank you for the clarifications

**Comment by morawi** [comment]
> As a differentiable operation, apart from retaining the computational graph, it will also provide meaningful gradients

I could be wrong, but, can the same be achieved using `retain_grad()`  for `torchvision` transformations? 
Which means, for the above example, `out` learns to approach `y`. No?

**Comment by shijianjian** [comment]
> > As a differentiable operation, apart from retaining the computational graph, it will also provide meaningful gradients
> 
> 
> 
> I could be wrong, but, can the same be achieved using `retain_grad()`  for `torchvision` transformations? 

It is a different concept. retain_grad aims on retaining the gradients that has been passed through the computational graph. While, what we mean a differentiable operation means to trace an operation.

You may refer to the logic that why STN is invented.


> Which means, for the above example, `out` learns to approach `y`, and `tr `changes accordingly. No?
> 
If 'tr' is an optimizable operation, then you are right about "changes accordingly".
